[#2716] include co-sign identifier (bsn) in submission report

pi-sigma opened the pull request:
Closes #2716 

* the identifier (bsn) of form co-signers is included in the submission report, alongside first + last name

CharString commented on src/openforms/submissions/tests/test_tasks_pdf.py:
```diff
@@ -243,4 +245,74 @@ def test_cosign_data_included(self):
+        )
+        self.assertInHTML(expected_identifier, rendered, count=1)
+
+    def test_incomplete_cosign_data(self):
```
I would split this into two tests... But first check with @joeribekker or @LaurensBurger if this is what the user story describes. I read the ticket as: require a BSN and show it + name if we know it.

pi-sigma commented:
A few tests needed refactoring after the changes. Most are straightforward, but there's one about which I'm not entirely sure: `test_co_sign_flow` in [src/openforms/authentication/tests/test_views](https://github.com/open-formulieren/open-forms/pull/2932/files#diff-c21b699bcaa2edeb83aac742a4e940ea88a4ff0a813f6cf58502d7403cfedb1d).

In order to make the test pass, I changed the `provides_auth` attribute of `authentication.tests.mocks.Plugin` to "bsn" instead of "dummy". I also removed the mock. The test passes with or without it, and I didn't see the point of it.

sergei-maertens commented on src/openforms/submissions/report.py:
```diff
@@ -15,6 +16,35 @@
+    def identifier(self) -> str:
+        return self.co_sign_data.get("identifier", "")
+
+    @property
+    def auth_attribute(self) -> str:
+        return self.co_sign_data.get("co_sign_auth_attribute", "")
```
this should use the human readable label of the `AuthAttributes` choices instead of the raw string value:

```
class AuthAttribute(models.TextChoices):
    """
    core identifying attributes retrieved from the authentication plugins
    """

    bsn = "bsn", _("BSN")
    kvk = "kvk", _("KvK number")
    pseudo = "pseudo", _("Pseudo ID")
    employee_id = "employee_id", _("Employee ID")
```

there's some sort of `AuthAttribute.labels` helper available, check the django docs :)

sergei-maertens commented on src/openforms/submissions/report.py:
```diff
@@ -15,6 +16,35 @@
+class CoSigner:
+    co_sign_data: dict
+
+    @property
+    def representation(self) -> str:
+        return self.co_sign_data.get("representation", "")
```
in general the patern `self.co_sign_data.get("...") or ""` is better to guarantee the type is correct if the key is absent or empty-ish (i.e. present but the value is `None`)

maybe a better approach would be a complete dataclass and have the `Report.co_signer` perform this normalization

sergei-maertens commented on src/openforms/submissions/report.py:
```diff
@@ -35,11 +65,13 @@ def show_payment_info(self) -> bool:
+        if not (co_sign_data := self.submission.co_sign_data):
+            return None
+
+        co_signer = CoSigner(co_sign_data)
```
this should be something more along the lines of:

```py
co_signer = CoSigner(identifier=co_sign_data.get('identifier") or "", representation=co_sign_data.get("representatin") or "", ...)
```
-> do the normalization in this central entrypoint to make the dataclass strongly and completely typed instead of dealing with flexible types like dicts inside of them.

sergei-maertens commented on src/openforms/submissions/report.py:
```diff
@@ -15,6 +16,35 @@
+    def auth_attribute(self) -> str:
+        return self.co_sign_data.get("co_sign_auth_attribute", "")
+
+    def __str__(self):
+        if not self.identifier:
+            return _(
+                "{representation} (missing {auth_attribute} for co-signer)"
+            ).format(
+                representation=self.representation, auth_attribute=self.auth_attribute
+            )
+        return f"{self.representation} ({self.auth_attribute}: {self.identifier})"
+
+    def __repr__(self):
+        return f"CoSigner(co_sign_data={self.co_sign_data}, auth_attribute={self.auth_attribute})"
```
these are typically generated by the dataclass and there should not be a need to override them.

If you need a custom/particular string representation, implement them as a custom method and call that in the template

pi-sigma commented:
I got rid of the dataclass for the co-signer, after all (it's not used elsewhere).

sergei-maertens commented on src/openforms/submissions/tests/test_tasks_pdf.py:
```diff
@@ -222,25 +222,56 @@ def test_cosign_data_included(self):
             <div class="submission-step-row">
                 <div class="submission-step-row__label">{key}</div>
-                <div class="submission-step-row__value">T. Shikari</div>
+                <div class="submission-step-row__value">T. Shikari (BSN: 123456782)</div>
             </div>
             """,
             key=_("Co-signed by"),
```
this value should also use the `gettext(...)`call otherwise if the shape changes in translations, this test will fail

sergei-maertens commented on src/openforms/submissions/serializers.py:
```diff
@@ -21,3 +23,6 @@ class CoSignDataSerializer(serializers.Serializer):
     def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
         self.fields["plugin"].choices = register.get_choices()
+
+    # def get_co_sign_auth_attribute(self, obj):
+    #     return obj.get_co_sign_auth_attribute_display()
```
```suggestion
```

sergei-maertens commented:
Backported via 050db5356252d5624f4209c16cbbefd4dec9aded
